src/App.js

```diff
@@ -1,13 +1,126 @@
+    } else {
+      const task = {
+        // slice creates a copy of this.state.input
+        task: this.state.input.slice(),
+        id: Date.now(),
+        completed: false
+      };
+
+      // copy todos and add task to todos
+      const todos = [...this.state.todos, task];
+
+      // add task to todo list
+      // todos.push(task);
+
+      this.setState({ todos: todos, input: '' });
+
+      // Local Storage
+      localStorage.setItem('todos', JSON.stringify(todos));
+      localStorage.setItem('input', '');
+    }
+  };
+
+  handleInputChange = e => {
+    // if warning message is showing - hide it!
+    // eslint-disable-next-line
+    this.state.invalidInput ? this.setState({ invalidInput: false }) : null;
+    this.setState({ input: e.target.value });
+    localStorage.setItem('input', e.target.value);
+  };
+
+  // map over existing array with conditional
+  toggleCompleteTask = e => {
+    const newTodos = this.state.todos.map(todo => {
+      if (String(todo.id) === e.target.id) {
+        return {
+          ...todo,
+          completed: !todo.completed
+        };
+      } else {
+        return todo;
+      }
+    });
+    this.setState({ todos: newTodos });
+    localStorage.setItem('todos', JSON.stringify(newTodos));
+  };
+
+  handleCompletedTasks = e => {
+    e.preventDefault();
+    const removeCompleted = this.state.todos.filter(todo => !todo.completed);
+    this.setState({ todos: removeCompleted });
+
+    // Local Storage
+    localStorage.setItem('todos', JSON.stringify(removeCompleted));
+  };
+
+  // retrive todos from local storage
+  hydrateStateWithLocalStorage() {
+    // for all items in state
+    for (let key in this.state) {
+      // if the key exists in localStorage
+      if (localStorage.hasOwnProperty(key)) {
+        // get the key's value from localStorage
+        let value = localStorage.getItem(key);
+        // parse the localStorage string and setState
+        try {
+          value = JSON.parse(value);
+          this.setState({ [key]: value });
+        } catch (e) {
+          // handle empty string
+          this.setState({ [key]: value });
+        }
+      }
+    }
+  }
+
+  // React.Component method
+  componentDidMount() {
+    this.hydrateStateWithLocalStorage();
+  }
+
   render() {
     return (
-      <div>
-        <h2>Welcome to your Todo App!</h2>
+      <div className="App">
+        <TodoHeader />
+        <TodoForm
+          handleNewToDo={this.handleNewToDo}
+          handleInputChange={this.handleInputChange}
+          handleCompletedTasks={this.handleCompletedTasks}
+          input={this.state.input}
+        />
+        <TodoMessage
+          inlineStyle={this.state.invalidInput ? this.state.showMessage : null}
+        />
+        <TodoList
+          todos={this.state.todos}
+          toggleCompleteTask={this.toggleCompleteTask}
+          style={this.state.todoStyle}
+          inlineStyle={this.state.toDoCompleteStyle}
```

Comment:
I wouldn't put this on state... it would not be something that was fluctuating. If you want to have an inlineStyle prop on your TodoList to make the representation flexible, just pass that as a prop with a string, or just have some sort of variable or import that you are keeping that theming info with.

But in my opinion CSS already handles this, you use the React and classnames to handle the "business logic" and you use the CSS to define what that actually looks like on the screen. So that's my preferred approach.

Since you're exploring patterns like this, I would say you will likely find styled-components very compelling when we learn about those, and other CSS-in-JS options.

I would say in any case don't put it on state though, state is for things that dynamically update. What dynamically updates is whether something is in an error state, or whether it's completed, not *how* that is visually represented.